lib/encoder: BenchmarkOneDriveREplaceNew fails

<!--

We understand you are having a problem with rclone; we want to help you with that!

**STOP and READ**
**YOUR POST WILL BE REMOVED IF IT IS LOW QUALITY**:
Please show the effort you've put in to solving the problem and please be specific.
People are volunteering their time to help! Low effort posts are not likely to get good answers!

If you think you might have found a bug, try to replicate it with the latest beta (or stable).
The update instructions are available at https://rclone.org/commands/rclone_selfupdate/

If you can still replicate it or just got a question then please use the rclone forum:

    https://forum.rclone.org/

for a quick response instead of filing an issue on this repo.

If nothing else helps, then please fill in the info below which helps us help you.

**DO NOT REDACT** any information except passwords/keys/personal info.

You should use 3 backticks to begin and end your paste to make it readable.

Make sure to include a log obtained with '-vv'.

You can also use '-vv --log-file bug.log' and a service such as https://pastebin.com or https://gist.github.com/

Thank you

The Rclone Developers

-->



#### What is the problem you are having with rclone?
I tried to run benchmarks on rclone but they failed instead with
```shell
$ go test -run=^$ -bench=.
--- FAIL: BenchmarkOneDriveReplaceNew
    encoder_test.go:290: Encode("\\*<>?:|#%\".~/\\*<>?:|#%\".~") want "＼＊＜＞？：｜＃％＂.~/＼＊＜＞？：｜＃％＂.~" got "＼＊＜＞？：｜＃％＂.~／＼＊＜＞？：｜＃％＂.~"
    encoder_test.go:290: Encode(" leading space/ leading space/ leading space") want " leading space/ leading space/ leading space" got " leading space／ leading space／ leading space"
    encoder_test.go:290: Encode("~leading tilde/~leading tilde/~leading tilde") want "～leading tilde/～leading tilde/～leading tilde" got "～leading tilde／~leading tilde／~leading tilde"
    encoder_test.go:290: Encode("leading tilde/~leading tilde") want "leading tilde/～leading tilde" got "leading tilde／~leading tilde"
    encoder_test.go:290: Encode("trailing dot./trailing dot./trailing dot.") want "trailing dot．/trailing dot．/trailing dot．" got "trailing dot.／trailing dot.／trailing dot．"
goos: darwin
goarch: amd64
pkg: github.com/rclone/rclone/lib/encoder
cpu: Intel(R) Core(TM) i7-7920HQ CPU @ 3.10GHz
BenchmarkOneDriveReplaceOld-8   	   46914	     24894 ns/op
--- FAIL: BenchmarkOneDriveRestoreNew
    encoder_test.go:301: Decode("~leading tilde/～leading tilde/～leading tilde") want "~leading tilde/~leading tilde/~leading tilde" got "~leading tilde/～leading tilde/～leading tilde"
    encoder_test.go:301: Decode("leading tilde/～leading tilde") want "leading tilde/~leading tilde" got "leading tilde/～leading tilde"
    encoder_test.go:301: Decode("trailing dot．/trailing dot．/trailing dot.") want "trailing dot./trailing dot./trailing dot." got "trailing dot．/trailing dot．/trailing dot."
BenchmarkOneDriveRestoreOld-8   	  295726	      4191 ns/op
FAIL
exit status 1
FAIL	github.com/rclone/rclone/lib/encoder	2.838s
```

and I discovered this by trying to add continuous benchmarking using Orijtech Inc's product ["Bencher"](https://bencher.orijtech.com/)  which produced the following https://dashboard.github.orijtech.com/benchmark/6eb7fed63a004f859bcea067383a9544
![Screen Shot 2021-10-04 at 1 06 04 AM](https://user-images.githubusercontent.com/4898263/135815511-bb64e206-211f-42e1-9e9a-dda29e7bc272.png)

and this can be reproduced on other systems. I highly encourage adding rclone for FREE to gobencher via the Github marketplace per https://github.com/marketplace/gobencher; every PR/commit will get benchmarked and you'll see what regressed or not 



#### What is your rclone version (output from `rclone version`)
Git commit 60323dc5e2fd0c4b269c1316784bf14adfacf5a6


#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
Both MacOS and Linux 64 bit


#### Which cloud storage system are you using? (e.g. Google Drive)
Not Applicable


#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
`go test -run=$^ -bench=. ./...`


#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)



<!--- Please keep the note below for others who read your bug report. -->

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

/cc-ing my colleagues @cuonglm @kirbyquerby

--- a/lib/encoder/encoder_test.go
+++ b/lib/encoder/encoder_test.go
@@ -266,54 +266,111 @@ const oneDrive = (Standard |
 	EncodeRightPeriod)
 
 var benchTests = []struct {
-	in  string
-	out string
+	in     string
+	outOld string
+	outNew string
 }{
-	{"", ""},
-	{"abc 123", "abc 123"},
-	{`\*<>?:|#%".~`, `＼＊＜＞？：｜＃％＂.~`},
-	{`\*<>?:|#%".~/\*<>?:|#%".~`, `＼＊＜＞？：｜＃％＂.~/＼＊＜＞？：｜＃％＂.~`},
-	{" leading space", " leading space"},
-	{"~leading tilde", "～leading tilde"},
-	{"trailing dot.", "trailing dot．"},
-	{" leading space/ leading space/ leading space", " leading space/ leading space/ leading space"},
-	{"~leading tilde/~leading tilde/~leading tilde", "～leading tilde/～leading tilde/～leading tilde"},
-	{"leading tilde/~leading tilde", "leading tilde/～leading tilde"},
-	{"trailing dot./trailing dot./trailing dot.", "trailing dot．/trailing dot．/trailing dot．"},
+	{
+		"",
+		"",
+		"",
+	},
+	{
+		"abc 123",
+		"abc 123",
+		"abc 123",
+	},
+	{
+		`\*<>?:|#%".~`,
+		`＼＊＜＞？：｜＃％＂.~`,
+		`＼＊＜＞？：｜＃％＂.~`,
+	},
+	{
+		`\*<>?:|#%".~/\*<>?:|#%".~`,
+		`＼＊＜＞？：｜＃％＂.~/＼＊＜＞？：｜＃％＂.~`,
+		`＼＊＜＞？：｜＃％＂.~／＼＊＜＞？：｜＃％＂.~`,
+	},
+	{
+		" leading space",
+		" leading space",
+		" leading space",
+	},
+	{
+		"~leading tilde",
+		"～leading tilde",
+		"～leading tilde",
+	},
+	{
+		"trailing dot.",
+		"trailing dot．",
+		"trailing dot．",
+	},
+	{
+		" leading space/ leading space/ leading space",
+		" leading space/ leading space/ leading space",
+		" leading space／ leading space／ leading space",
+	},
+	{
+		"~leading tilde/~leading tilde/~leading tilde",
+		"～leading tilde/～leading tilde/～leading tilde",
+		"～leading tilde／~leading tilde／~leading tilde",
+	},
+	{
+		"leading tilde/~leading tilde",
+		"leading tilde/～leading tilde",
+		"leading tilde／~leading tilde",
+	},
+	{
+		"trailing dot./trailing dot./trailing dot.",
+		"trailing dot．/trailing dot．/trailing dot．",
+		"trailing dot.／trailing dot.／trailing dot．",
+	},
 }
 
-func benchReplace(b *testing.B, f func(string) string) {
+func benchReplace(b *testing.B, f func(string) string, old bool) {
 	for range make([]struct{}, b.N) {
 		for _, test := range benchTests {
 			got := f(test.in)
-			if got != test.out {
-				b.Errorf("Encode(%q) want %q got %q", test.in, test.out, got)
+			out := test.outNew
+			if old {
+				out = test.outOld
+			}
+			if got != out {
+				b.Errorf("Encode(%q) want %q got %q", test.in, out, got)
 			}
 		}
 	}
 }
 
-func benchRestore(b *testing.B, f func(string) string) {
+func benchRestore(b *testing.B, f func(string) string, old bool) {
 	for range make([]struct{}, b.N) {
 		for _, test := range benchTests {
-			got := f(test.out)
+			out := test.outNew
+			if old {
+				out = test.outOld
+			}
+			got := f(out)
 			if got != test.in {
-				b.Errorf("Decode(%q) want %q got %q", got, test.in, got)
+				b.Errorf("Decode(%q) want %q got %q", out, test.in, got)
 			}
 		}
 	}
 }
+
 func BenchmarkOneDriveReplaceNew(b *testing.B) {
-	benchReplace(b, oneDrive.Encode)
+	benchReplace(b, oneDrive.Encode, false)
 }
+
 func BenchmarkOneDriveReplaceOld(b *testing.B) {
-	benchReplace(b, replaceReservedChars)
+	benchReplace(b, replaceReservedChars, true)
 }
+
 func BenchmarkOneDriveRestoreNew(b *testing.B) {
-	benchRestore(b, oneDrive.Decode)
+	benchRestore(b, oneDrive.Decode, false)
 }
+
 func BenchmarkOneDriveRestoreOld(b *testing.B) {
-	benchRestore(b, restoreReservedChars)
+	benchRestore(b, restoreReservedChars, true)
 }
 
 var (

